src/modules/analytics/withAnalytics.js

```diff
@@ -1,141 +1,208 @@
+            }
+            const boundCallback = (...args) => {
+              const event = this.getEnqueuedEvent(eventName);
+              providedCallback(...args, event);
+            };
+            return {
+              ...bound,
+              [propName]: boundCallback,
+            };
+          }, {});
+
+        return (({ ...props, ...boundPropCallbacks }: any): Props);
       }
 
-      return Object.keys(bindEvents)
-        .reduce((bound, eventName) => {
-          const propName = bindEvents[eventName];
-          const providedCallback = this.props[propName];
-          if (!providedCallback) {
-            return bound;
-          }
-          const boundCallback = (...args) => {
-            const event = this.getEnqueuedEvent(eventName);
-            providedCallback(...args, event);
-          };
-          return {
-            ...bound,
-            [propName]: boundCallback,
-          };
-        }, {});
-    }
-
-    render() {
-      const {
-        analyticsNamespace,
-        ...permittedProps
-      } = this.props;
-
-      const props = {
-        ...permittedProps,
-        ...this.bindEventsToProps(),
-        createAnalyticsEvent: this.createAnalyticsEvent,
-      };
+      // Consumers can provide a map of callbacks to event creators to allow creating events
+      // within certain components.
+        mapCreateEventsToProps = (props: Props): Props => {
+        if (!createEventMap) {
+          return props;
+        }
+
+        const modifiedPropCallbacks = Object.keys(createEventMap)
+          .reduce((modified, propCallbackName) => {
+            const eventCreator = createEventMap[propCallbackName];
+            const providedCallback = props[propCallbackName];
+            if (!providedCallback) {
+              return modified;
+            }
+            const modifiedCallback = (...args) => {
+              if (typeof eventCreator === 'string') {
+                this.createAnalyticsEvent(eventCreator).raise();
+              } else if (typeof eventCreator === 'function') {
+                eventCreator(this.createAnalyticsEvent, props).raise();
+              } else {
+                console.error('Invalid event creator type passed to withAnalytics event map');
+              }
+              providedCallback(...args);
+            };
+            return {
+              ...modified,
+              [propCallbackName]: modifiedCallback,
+            };
+          }, {});
+
+        return (({ ...props, ...modifiedPropCallbacks }: any): Props);
+      }
 
-      return (
-        <WrappedComponent {...props} />
-      );
-    }
-  };
+      render() {
+        const patchedProps = this.mapCreateEventsToProps(this.bindEventsToProps(this.props));
+
+        const {
+          analyticsNamespace,
+          analytics,
+          bindEventsToProps,
+          ...permittedProps
+        } = patchedProps;
+
+        const props = {
+          ...permittedProps,
+          ...patchedProps,
```

Comment:
i don't think we should be merging `patchedProps` here, just `permittedProps` right?

Reply:
Yeah, good catch.